src/post-submission-actions/program-enrollment-action.ts

```diff
@@ -0,0 +1,97 @@
+import { showToast } from '@openmrs/esm-framework';
+import { createProgramEnrollment, getPatientEnrolledPrograms, updateProgramEnrollment } from '../api/api';
+import { PostSubmissionAction } from '../api/types';
+import dayjs from 'dayjs';
+
+export const ProgramEnrollmentSubmissionAction: PostSubmissionAction = {
+  applyAction: async function ({ patient, encounters, sessionMode }, config) {
+    const encounter = encounters[0];
+    const encounterLocation = encounter.location['uuid'];
+    // only do this in enter or edit mode.
+    if (sessionMode === 'view') {
+      return;
+    }
+
+    const enrollmentDate = encounter.obs?.find((item) => item.formFieldPath.includes(config.enrollmentDate))?.value;
+    const completionDate = encounter.obs?.find((item) => item.formFieldPath.includes(config.completionDate))?.value;
+    const programUuid = config.programUuid;
+
+    if (programUuid) {
+      const abortController = new AbortController();
+      const payload = {
+        patient: patient.id,
+        program: programUuid,
+        dateEnrolled: enrollmentDate ? dayjs(enrollmentDate).format() : null,
+        dateCompleted: completionDate ? dayjs(enrollmentDate).format() : null,
+        location: encounterLocation,
+      };
+
+      if (sessionMode === 'enter') {
+        const patientEnrolledPrograms = await getPatientEnrolledPrograms(patient.id);
+        if (patientEnrolledPrograms) {
+          const hasActiveEnrollment = patientEnrolledPrograms.results.some(
+            (enrollment) => enrollment.program.uuid === programUuid && enrollment.dateCompleted === null,
+          );
+          if (hasActiveEnrollment) {
+            throw new Error('Cannot enroll patient to program. Patient already has an active enrollment');
+          }
+        }
+        createProgramEnrollment(payload, abortController).subscribe(
+          (response) => {
+            if (response.status === 201) {
+              showToast({
+                critical: true,
+                kind: 'success',
+                description: 'It is now visible in the Programs table',
+                title: 'Program enrollment saved',
+              });
+            }
+          },
+          (err) => {
+            showToast({
+              title: 'Error saving program enrollment',
+              kind: 'error',
+              critical: false,
+              description: err?.message,
+            });
+          },
+        );
+      } else {
+        const patientEnrolledPrograms = await getPatientEnrolledPrograms(patient.id);
```

Comment:
`if (hasActiveEnrollment) {
        throw new Error('Cannot enroll patient to program. Patient already has an active enrollment');
 }`
Will this section get executed because you are checking if someone is already enrolled ?

Reply:
Not sure I fully get the question but let me briefly explain what is happening here;
So there's an existing module through which patients are being enrolled into programs. However we are also adding provision for enrolling patients into programs after successful submission of some encounters using post submission actions. Now when the post submission action is in `enter mode`, the assumption is that someone is being enrolled into a program for the 1st time, however if they already has an active program of the same type, they can't be enrolled again. This is being handled by the backend but saw it wise to handle it on the frontend as well. And yah, the section will get executed.